code/parse/sexp.cpp

```diff
@@ -29803,6 +30945,222 @@ int check_text_for_variable_name(const char *text)
+
+		if (value_it != container.map_data.end()) {
+			result = value_it->second;
+			return true;
+		} else {
+			if (Log_event) {
+				SCP_string log_string = container.container_name;
+				log_string.append(" map container returned nothing when searched for key ");
+				log_string.append(key);
+				Current_event_log_container_buffer->emplace_back(log_string);
+			}
+			return false;
+ 		}
+	} else if (container.is_list()) {
+		if (container.list_data.empty()) {
+			Warning(LOCATION, "Attempt to retrieve data from empty list container %s", container.container_name.c_str());
+			return false;
+		}
+
+		const char *modifier = CTEXT(node); 
+
+		int modifier_index = 0;
+		for (; modifier_index < NUM_CONTAINER_MODIFIERS; modifier_index++) {
+			if (!stricmp(modifier, Container_modifiers[modifier_index].name)) {
+				break;
+			}
+		}
+
+		if (modifier_index == NUM_CONTAINER_MODIFIERS) {
+			const SCP_string &container_name = container.container_name;
+			Warning(LOCATION,
+				"Illegal operation attempted on %s container. There is no modifier called %s.",
+				container_name.c_str(),
+				modifier);
+			log_printf(LOGFILE_EVENT_LOG,
+				"Illegal operation attempted on %s container. There is no modifier called %s.",
+				container_name.c_str(),
+				modifier);
+			return false;
+		}
+
+		int data_index;
+		auto &list_data = container.list_data;
+		auto list_it = list_data.begin();
+
+		switch (modifier_index) {
+			case SNF_CONTAINER_GET_FIRST:
+				result = list_data.front();
+				return true;
+
+			case SNF_CONTAINER_GET_LAST:
+				result = list_data.back();
+				return true;
+
+			case SNF_CONTAINER_REMOVE_FIRST:
+				result = list_data.front();
+				if (are_containers_modifiable()) {
+					list_data.pop_front();
+				}
+				return true;
+
+			case SNF_CONTAINER_REMOVE_LAST:
+				result = list_data.back();
+				if (are_containers_modifiable()) {
+					list_data.pop_back();
+				}
+				return true;
+
+			case SNF_CONTAINER_GET_RANDOM:
+				data_index = rand_internal(0, (int)list_data.size() - 1);
+				result = *std::next(list_it, data_index);
+				return true;
+
+			case SNF_CONTAINER_REMOVE_RANDOM:
+				data_index = rand_internal(0, (int)list_data.size() -1);
+				std::advance(list_it, data_index);
+				result = *list_it;
+				if (are_containers_modifiable()) {
+					list_data.erase(list_it);
+				}
+				return true;
+
+			case SNF_CONTAINER_AT_INDEX:
+				node = CDR(node);
+				if (node == -1) {
+					return false;
+				}
+				bool is_nan, is_nan_forever;
+				data_index = eval_num(node, is_nan, is_nan_forever); 
+				if (is_nan || is_nan_forever) {
+					return false;
+				}
+				Assert(data_index >= 0);
+				Assert(data_index < (int)list_data.size());
+				result = *std::next(list_it, data_index);
+				return true;
+
+			default:
+				Error(LOCATION, "Unknown modifier type found in container"); 
+				break;
+		}
+	}
+
+	return false;
+}
+
+int container_push_return_string(const SCP_string &result)
```

Comment:
This part of the PR is probably the one I'm the most concerned about.

When containers are used with `CTEXT()`, we store the resulting strings in a global array with 🤞🏻 that the string won't be overwritten throughout the time that we need it.

Maybe the resulting string should be written into the `Sexp_node`'s cached data instead?

Reply:
I agree with the concern, but what could be overwriting it?